Object replacement character (0xFFFC) generates null (0) glyph

litherum opened the pull request:
#### aba48e65ac8b5a624f37a9be944e686e552d89b7
<pre>
Object replacement character (0xFFFC) generates null (0) glyph
<a href="https://bugs.webkit.org/show_bug.cgi?id=255298">https://bugs.webkit.org/show_bug.cgi?id=255298</a>
rdar://107897203

Reviewed by Cameron McCormack.

The spec doesn&apos;t say anything about the object replacement character being deleted. Neither
Chrome nor Firefox delete the object replacement character.

Tests are being upstreamed at <a href="https://github.com/web-platform-tests/wpt/pull/40391.">https://github.com/web-platform-tests/wpt/pull/40391.</a>

* LayoutTests/TestExpectations:
* LayoutTests/fast/text/zero-width-characters-complex-script.html:
* LayoutTests/fast/text/zero-width-characters.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-combine-emphasis-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-combine-emphasis.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-1-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-1.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-2-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-2.html: Added.
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering):
(WebCore::FontCascade::treatAsZeroWidthSpaceInComplexScript):

Canonical link: <a href="https://commits.webkit.org/264886@main">https://commits.webkit.org/264886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1af77888ce5ed2e2e913f73e513bd3e0108bbb29

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/9306 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9525 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10672 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8981 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9027 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/11293 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/9274 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11835 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9165 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/10141 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7906 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10831 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7436 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/8242 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/15727 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8545 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/8390 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11717 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8890 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/7249 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8134 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/8149 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2176 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12349 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8640 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on LayoutTests/fast/text/text-combine-emphasis-expected-mismatch.html:
```diff
@@ -0,0 +1,16 @@
+.foo {
+  font-size: 100px;
+  writing-mode:vertical-lr;
+  text-combine-upright:all;
```
Nit: spaces after colons, for consistency.

heycam commented on LayoutTests/fast/text/text-combine-emphasis.html:
```diff
@@ -0,0 +1,17 @@
+  font-size: 100px;
+  writing-mode:vertical-lr;
+  text-combine-upright:all;
+  text-emphasis:filled;
```
And here.

heycam reviewed (approved):
r=me but consider making the new tests WPTs.

litherum committed:
Object replacement character (0xFFFC) generates null (0) glyph
https://bugs.webkit.org/show_bug.cgi?id=255298
rdar://107897203

Reviewed by Cameron McCormack.

The spec doesn't say anything about the object replacement character being deleted. Neither
Chrome nor Firefox delete the object replacement character.

Tests are being upstreamed at https://github.com/web-platform-tests/wpt/pull/40391.

* LayoutTests/TestExpectations:
* LayoutTests/fast/text/zero-width-characters-complex-script.html:
* LayoutTests/fast/text/zero-width-characters.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-combine-emphasis-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text-decor/text-combine-emphasis.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-1-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-1.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-2-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/white-space/object-replacement-2.html: Added.
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::isCharacterWhoseGlyphsShouldBeDeletedForTextRendering):
(WebCore::FontCascade::treatAsZeroWidthSpaceInComplexScript):

Canonical link: https://commits.webkit.org/264886@main
